### PR TITLE
MTDSA-11667 Trigger a self-assessment tax calculation - controller

### PR DESCRIPTION
**Pull request by mattstephens-hmrc**

### CODE REVIEW

File: test/v3/controllers/TriggerCalculationControllerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+          status(result) shouldBe expectedStatus
+          contentAsJson(result) shouldBe Json.toJson(error)
+          header("X-CorrelationId", result) shouldBe Some(correlationId)
+
+          MockedAuditService.verifyAuditEvent(auditError(expectedStatus, error)).once
+        }
+      }
+
+      val input = Seq(
```
Comment (vfryer12) [inline test/v3/controllers/TriggerCalculationControllerSpec.scala:209]: Shouldn't the nino and taxYear format errors be included here?

File: test/v3/controllers/TriggerCalculationControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        (RuleNoIncomeSubmissionsExistError, FORBIDDEN),
+        (RuleFinalDeclarationReceivedError, FORBIDDEN),
+        (BadRequestError, INTERNAL_SERVER_ERROR),
```
Comment (vfryer12) [inline test/v3/controllers/TriggerCalculationControllerSpec.scala:212]: BadRequestError should be mapped to BAD_REQUEST

File: it/v3/endpoints/TriggerCalculationControllerISpec.scala [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+        val input = Seq(
+          (BAD_REQUEST, "INVALID_NINO", BAD_REQUEST, NinoFormatError),
+          (BAD_REQUEST, "INVALID_TAX_YEAR", BAD_REQUEST, TaxYearFormatError),
+          (FORBIDDEN, "INVALID_TAX_CRYSTALLISE", BAD_REQUEST, FinalDeclarationFormatError),
+          (FORBIDDEN, "NO_SUBMISSION_EXIST", FORBIDDEN, RuleNoIncomeSubmissionsExistError),
+          (CONFLICT, "CONFLICT", FORBIDDEN, RuleFinalDeclarationReceivedError),
+          (INTERNAL_SERVER_ERROR, "SERVER_ERROR", INTERNAL_SERVER_ERROR, DownstreamError),
+          (SERVICE_UNAVAILABLE, "SERVICE_UNAVAILABLE", INTERNAL_SERVER_ERROR, DownstreamError)
+        )
```
Comment (vfryer12) [inline it/v3/endpoints/TriggerCalculationControllerISpec.scala:155]: INVALID_REQUEST error is missing

File: it/v3/endpoints/TriggerCalculationControllerISpec.scala [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+          | "calculationId" : "f2fb30e5-4ab6-4a29-b3c1-c7264259ff1c",
+          | "links":[
+          |   {
+          |   "href":"/individuals/calculations/AA123456A/self-assessment/2018-19?finalDeclaration=true",
+          |   "method":"POST",
+          |   "rel":"trigger"
+          |     }
+          |   ]
+          |}
+        """.stripMargin
```
Comment (vfryer12) [inline it/v3/endpoints/TriggerCalculationControllerISpec.scala:73]: Shouldn't the hateaos include the Retrieve links implementation as well?

File: app/v3/controllers/TriggerCalculationController.scala [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+  private def errorResult(errorWrapper: ErrorWrapper) = {
+    errorWrapper.error match {
+      case NinoFormatError | TaxYearFormatError                                  => BadRequest(Json.toJson(errorWrapper))
+      case RuleTaxYearNotSupportedError | RuleTaxYearRangeInvalidError           => BadRequest(Json.toJson(errorWrapper))
+      case FinalDeclarationFormatError                                           => BadRequest(Json.toJson(errorWrapper))
+      case RuleNoIncomeSubmissionsExistError | RuleFinalDeclarationReceivedError => Forbidden(Json.toJson(errorWrapper))
+
+      case BadRequestError => InternalServerError(Json.toJson(errorWrapper))
```
Comment (vfryer12) [inline app/v3/controllers/TriggerCalculationController.scala:116]: BadRequestError should be mapped to BadRequest(Json.toJson(errorWrapper)

File: it/v3/endpoints/TriggerCalculationControllerISpec.scala (hunk 4 above)
Comment (mattstephens-hmrc) [reply it/v3/endpoints/TriggerCalculationControllerISpec.scala:73]: the retrieve links are still in the separate feature branch I think

**Pull request merged**